### DISCUSSION TITLE
Is there a getting started -like guide?

### DISCUSSION THREAD

**Discussion by YJDoc2** [obi1kenobi/trustfall#318] [General]
Hello :wave:  Thanks a lot for making this amazing project! This seems something really interesting .

I was wondering is there any getting started guide that would basically do a walkthrough for writing an adapter and using this to create complex, cross data-source queries ; similar to the hackernews - github -yaml  example, but in a more tutorial style?

If that doesn't exist , I think it would be a great addition for beginners and interested people to get started with using this, and thinking of how this can be used for their own problems. One issue I'm facing with using this right now, even for a play-project ; is that because this is so new, and unlike anything else I have seen - multi data sources, with power to query from a single interface - it is hard for me to imagine exactly what I should use this for. This might be just a -me- problem, but I still think a tutorial like getting started guide would be a good addition if it doesn't exist yet.

Also is there any docs on the query language itself? I saw that it is supposed to be a subset of graphql, but having a single place to refer possible operators and filter and general syntax might be helpful for newcomers.

In case something like both above do not exist, would you be willing to accept PRs related to them/ discuss on how they should be done ( in repo itself, seprate mdbook, a single blog post somewhere or something else altogether)?

Thanks a lot!

**Comment by obi1kenobi** [comment]
The docs are currently minimal and woefully incomplete. I'd be thrilled to accept PRs to improve the README and crate doc-comments, and either add a docs directory to the repo or use Discussions here to post tutorials and getting started guides — not sure which would be better.

I think the problem you describe is very common, and very worth solving. I know this is a poor answer, but right now, the best source for learning the query language is going through the example queries in the playground ([hackernews](https://play.predr.ag/hackernews), [Rust APIs](https://play.predr.ag/rustdoc)), which cover nearly all of the language functionality.

If you're inclined to write some of the resources you described, I'd be thrilled to support you by answering questions, reading drafts, and anything else you'd need. This is a super ambitious project on which I'm stretched quite thin, so I'd greatly appreciate the help!

**Reply by YJDoc2** [reply]
Hey, Thanks for the reply! I'd be happy to try out writing some guides/docs that would solve the problems that I had mentioned. Currently my thoughts are this :
- Personally I'd prefer to have tutorial and docs as a document(s) in repo itself, if possible deployed as github pages or something. I think discussion posts might not be a good place for have the tutorials or guides ; just my personal thought.
- Whichever way we finally decide to create it, we should definitely link it in the readme, as front-and-center it can be. I re-read the readme 2/3 times thinking I was missing the docs link and checked the linked pages in case they have docs links. I think anyone stumbling on this either through repo or crates.io can benefit a lot if we have docs link directly there.
-   \* If * we do decide to go ahead with having tutorial in repo and not as community posts ; I think we can do it using mdbook, which seems to be pretty standard in rust eco-system, and can be easily deployed using gh-pages

As I said previously, even I'm not sure how can I start using trusfall in a project, and because I'm not much familiar with query syntax (and graphql as well) I'd like to start by writing  tutorial first, and maybe do some doc for query language as I get more familiar with it. For that I'm thinking :
- Have a very basic tutorial, which uses just a single data source, and shows basically how to set up trustfall, and how we can run queries and write an adapter , very basic.
- Then have a more complex one, which shows how we can have multiple data sources, but this would still be pretty simple comparatively. I read in one of your blog post (regarding how optimizing this improved semvar check by 2000x) that trustfall also gives query_info (which I haven't checked yet :sweat_smile: ) ; and I do not think that should go into this part.
- Then we can have maybe more advance tutorials on how to use query_info or so.

If thinking in terms of a blog, each of the above will be a separate post in a series. We can have a section in mdbook, in which each of this is a part. I think based on first two, newcomers should at least be able to see where trustfall can be used, and how to get started with it, hacking on their own sources.

I think using an example which would be familiar to a developer, and already has a solution, and then showing how trustfall can be used to solve it ; can help people get intuition about how trustfall fits in with rest of tools/solutions/scripts they might use.

---

Looking forward to hearing your thoughts on this! As I said, this is  just my personal opinion, so if you are thinking anything different, please let me know! Thanks :)

**Reply by YJDoc2** [reply]
Also a quick question -
I haven't looked at implementing an adapter yet apart from that it is a trait based thing ; but are trustfall adapters composable?

For example if someone has an adapter for github, and someone else has an adapter for yaml, can I add both packages as dependencies and then compose them, so I can write a single query which queries github raw files and then process ci yamls ? Not necessarily as simple as this, but essentially something like this. Or do we need to pipeline them manually, so I create query engine using one adapter, get its results, then create a new instance using second adapter and pipe the data from first into second and use two queries to get the final result?

Apologies for throwing a question without doing too much research :sweat:

**Reply by obi1kenobi** [reply]
Currently traveling, apologies for brevity. These are all reasonable
questions. Absolutely no need to apologize, it's on me for not making the
answers easier to find.

Adapters are designed to be composable, but I haven't yet published APIs
for that. It should be a fairly automatic process, but good APIs are hard.
Of course anyone could implement their own composition in the meantime.

Happy to include docs in the repo. Please feel free to act with agency in
designing and writing the docs -- you're in charge. I'm happy to merge PRs
and also happy to proofread, answer questions, or offer suggestions if
you'd like.

I'd recommend keeping the querying tutorial separate from the adapter
implementation tutorial, since the latter is specific to the programming
language (Rust, Python, JS/TS, etc.) while the query syntax is always the
same.

I also have some adapter autogeneration tools I plan to publish soon, which
should make adapter writing easier and should be part of those tutorials.

Thanks again! My schedule should be more back to normal in around a week or
so, and I hope to be able to help with docs and anything else you might
need.

**Reply by YJDoc2** [reply]
Hey, no problem!
Thanks for the response, I'll get working on the initial setup/basic guide as discussed. I'll figure out what would be a good example for a new user to write adapter, and get back to you when I have something ready and written :+1:

**Reply by obi1kenobi** [reply]
Feel free to send drafts my way as well. It doesn't have to be fully ready. "A little feedback early is better than a lot of feedback later" and all that :)

**Reply by YJDoc2** [reply]
Hey, yep! I was a bit busy last few days, so couldn't update much. I'll definitely get the feedback on drafts as well, not just the "final" version :) I'll update here once I have something to show, thanks!
